### DISCUSSION TITLE
Julia stuff

### DISCUSSION THREAD

**Discussion by jbcaillau** [control-toolbox/control-toolbox.github.io#22] [Ideas]
# Profiling
@PierreMartinon 
- [Performance tips](https://docs.julialang.org/en/v1/manual/performance-tips/)
- [Profiling (native)](https://web.mit.edu/julia_v0.6.2/julia/share/doc/julia/html/en/manual/profile.html)
- [Cthulhu](https://github.com/JuliaDebug/Cthulhu.jl)
- [BenchmarkTools](https://github.com/JuliaCI/BenchmarkTools.jl)
- [ProfileView](https://github.com/timholy/ProfileView.jl)

**Comment by ocots** [comment]
Docstrings extensions: 
* [doc of the package](https://docstringextensions.juliadocs.org/latest/)
* [need to be added into the Project.toml](https://github.com/SciML/SciMLBase.jl/blob/master/Project.toml)
* [example](https://github.com/SciML/SciMLBase.jl/blob/master/src/ensemble/basic_ensemble_solve.jl#L2)

For instance:
```julia
using DocStringExtensions

"""
A short summary of `func`...

$(SIGNATURES)

where `x` and `y` should both be positive.

# Details

Some details about `func`...
"""
func(x, y) = x + y
```

**Reply by ocots** [reply]
@joseph-gergaud @PierreMartinon If you want to write some docstrings, you can use `DocStringExtensions`, cf. above.

**Comment by jbcaillau** [comment]
https://info.juliahub.com/en-us/tools-for-non-linear-mpc

**Comment by jbcaillau** [comment]
Struct equality test:
https://discourse.julialang.org/t/surprising-struct-equality-test/4890/8

**Comment by jbcaillau** [comment]
@ocots any idea where has the wiki page with info on `registry add, dev...` moved?

**Reply by ocots** [reply]
https://github.com/control-toolbox/control-toolbox.github.io/wiki/Dev-setup

**Comment by jbcaillau** [comment]
Sone knows the point in using ``x_1`` instead of $x_1$ for Mardown math? See, *e.g.* [this source file](https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/blob/main/docs/src/tutorial.md)
and [this result file](https://juliasmoothoptimizers.github.io/ADNLPModels.jl/dev/tutorial).

**Comment by jbcaillau** [comment]
Nice Pluto notebooks (`.jl`, not `.ipynb`):

[Invariant manifolds](https://github.com/cadojo/CR3BP-Manifold-Research/blob/main/notebooks/manifold-transfers/Invariant%20Manifolds/invariant-manifolds-in-space.jl) [^1]

[Going to Jupiter](https://github.com/cadojo/CR3BP-Manifold-Research/blob/main/notebooks/manifold-transfers/Going%20to%20Jupiter%20with%20Julia/Manifolds.jl)

[^1]: see [here](https://juliahub.com/ui/Notebooks/cadojo/Astrodynamics/invariant-manifolds-in-space.jl) for `Pluto` rendering on `JuliaHub`

**Comment by jbcaillau** [comment]
Problem with `conda` install of `julia`:
- missing `cert.pm` (@j-l-s rings a bell 🙁?)
- problem with package install (`HTTP`, `Plotly`...)

AVOID, and use official release instead; see [this post](https://discourse.julialang.org/t/problems-installing-packages-cert-pm-no-such-file/83995/8). Everything back to normal with `1.8.5` local install.

**Comment by ocots** [comment]
Taylor integration : https://docs.sciml.ai/TaylorIntegration/stable/common/

**Comment by ocots** [comment]
Doc and nice benchmark: https://juliasmoothoptimizers.github.io/Percival.jl/dev/

**Comment by jbcaillau** [comment]
# On julia vs. markdown vs. notebooks files

[Weave.jl](https://weavejl.mpastell.com)

[Literate.jl](https://fredrikekre.github.io/Literate.jl/v2/)

**Comment by jbcaillau** [comment]
# 20 Awesome Julia Tips And Tricks For The Advantageous Programmer
[Check this post](https://towardsdatascience.com/20-awesome-julia-tips-and-tricks-for-the-advantageous-programmer-5fedbfd00f73)

**Comment by j-l-s** [comment]
How to create "real" getters and setters (№17: “ object-oriented programming”)

**Reply by jbcaillau** [reply]
On no. 17: too convoluted. Julia is not o-oriented, dot!

**Reply by rveltz** [reply]
getters and setters are a must for big projects. Check Kwong, Tom. Hands-on Design Patterns and Best Practices with Julia: Proven Solutions to Common Problems in Software Design for Julia 1.x. Birmingham, UK: Packt Publishing, 2020.

**Comment by PierreMartinon** [comment]
A very long post indeed. I struggled to find any useful information in it, but that's probably me 😄
  

  

  

  
  
  
  
>

**Comment by jbcaillau** [comment]
# Traits & more

https://github.com/JuliaLang/julia/issues/2345#issuecomment-54537633

https://invenia.github.io/blog/2019/10/30/julialang-features-part-1/

https://invenia.github.io/blog/2019/11/06/julialang-features-part-2/

https://github.com/mauro3/SimpleTraits.jl

https://github.com/jolin-io/WhereTraits.jl

https://github.com/PacktPublishing/Hands-on-Design-Patterns-and-Best-Practices-with-Julia

http://ucidatascienceinitiative.github.io/IntroToJulia/Html/DispatchDesigns

https://docs.julialang.org/en/v1/manual/methods/#Trait-based-dispatch-1

https://youtu.be/qhAe1PuWjRQ?si=9JWFxqvSfRDAUvO6

## See also
[Traits in rust](https://en.wikipedia.org/wiki/Rust_(programming_language)#Traits)

## Example
From https://www.oxinabox.net/2018/10/03/Dispatch,-Traits-and-Metaprogramming-Over-Reflection.html#part-3-traits-falling-back-to-hasmethod

```julia
# The trait types
abstract type Scalarness end
struct Scalar <: Scalarness end
struct NonScalar <: Scalarness end

# The trait functions (extensible! contrary to unions)
scalarness(::Type) = Scalar() #fall-back, by default everything is scalar; could 
scalarness(::Type{<:AbstractArray}) = NonScalar()
scalarness(::Type{<:Tuple}) = NonScalar()
scalarness(::Type{<:AbstractString}) = NonScalar()

# Dispatching statically (values are types, known statically - nb: also works on small values such as booleans...)
aslist(x::T) where T = aslist(scalarness(T), x)
aslist(::Scalar, x) = [x]
aslist(::NonScalar, x) = x
```

**Comment by jbcaillau** [comment]
[Perf. profiling in Julia](https://www.youtube.com/watch?v=pvduxLowpPY)

**Comment by jbcaillau** [comment]
@ocots *à l'occasion*, please write somewhere what you suggest to build a environment for students instead of them duplicating packages in their own `.julia`

**Comment by jbcaillau** [comment]
# Metaprogramming

[get-the-name-of-the-invoking-module](https://discourse.julialang.org/t/get-the-name-of-the-invoking-module/22685/15)

## Generated functions
https://docs.julialang.org/en/v1/manual/metaprogramming/#Generated-functions
https://discourse.julialang.org/t/how-do-i-create-a-function-from-an-expression/12349/23
https://github.com/SciML/RuntimeGeneratedFunctions.jl
https://github.com/JuliaStaging/GeneralizedGenerated.jl
[Base.@locals](https://discourse.julialang.org/t/problem-evaluating-symbol-for-macro-inside-function/30142/4)

```julia
function genfun2(x, y, body, gs=gensym()) # todo: macro for varargin
    eval(Expr(:function, Expr(:call, gs, x, y), body))
    (a, b) -> Base.invokelatest(eval(gs), a, b)
end

a = 1
f1 = genfun2(:x, :y, :(x + y + a))
f1(1, 2)

f2(t) = begin
    foo = genfun2(:x, :y, :(x + 2y + a))
    foo(1, t[1])
end
f2([ 2 ])

# compare (no invokelatest)
function _genfun2(x, y, body, gs=gensym())
    eval(Expr(:function, Expr(:call, gs, x, y), body))
end

a = 1
f3 = _genfun2(:x, :y, :(x + y + a))
f3(1, 2)

f4(t) = begin
    foo = _genfun2(:x, :y, :(x + y + a))
    println("locals: ", Base.@locals())
    foo(1, t)
end
#f4(2) # foo is not know in the global scope and needs to be retrieved thanks to invokelatest
```

**Comment by jbcaillau** [comment]
@joseph-gergaud @ocots @PierreMartinon worth a try!
> What’s New in Julia 1.9: Dr. Jeff Bezanson, Julia co-creator and JuliaHub co-founder and CTO, hosted a Webinar on Julia 1.9. Julia 1.9 fixes the longstanding TTFX, or Time to First Execution issue (often called Time to First Plot). Package precompilation now saves native code into a “package image”, accelerating the first plot operation to a fraction of a second. There are several other improvements to threading, packages, and compilation, with an exhaustive list in the Julia 1.9 release notes. Click here to watch.

**Comment by jbcaillau** [comment]
@ocots Julia guideline for doc: 
> The one-line sentence should use the imperative form ("Do this", "Return that") instead of the third person (do not write "Returns the length...") when documenting functions. 

See [here](https://docs.julialang.org/en/v1/manual/documentation/)

**Comment by ocots** [comment]
# Binary operators

- https://docs.julialang.org/en/v1/base/math/#Customizable-binary-operators
- https://github.com/JuliaLang/julia/blob/master/src/julia-parser.scm

**Comment by jbcaillau** [comment]
# Package compatibility

🙏🏽 @rveltz for the reco
[CompatHelper.jl](https://github.com/JuliaRegistries/CompatHelper.jl)

**Comment by jbcaillau** [comment]
Has someone tried Julia 1.9? (fixing TTFX, *a priori*)

**Reply by ocots** [reply]
Not yet.

**Reply by jbcaillau** [reply]
A quick try; `CTBase` not loading (cryptic load error). Retrying later.

**Comment by jbcaillau** [comment]
# Good practices

🙏🏽 @rveltz for the pointers

[ColPrac](https://github.com/SciML/ColPrac)

[SciML style guide](https://github.com/SciML/SciMLStyle)

**Comment by jbcaillau** [comment]
HSL lib for ipopt, [check this](https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/pull/127#issuecomment-1569573550)

**Comment by ocots** [comment]
[The dangers of abusing multiple dispatch (aka, more on types with values-as-parameters)](https://docs.julialang.org/en/v1/manual/performance-tips/#The-dangers-of-abusing-multiple-dispatch-(aka,-more-on-types-with-values-as-parameters))

- https://groups.google.com/g/julia-users/c/jUMu9A3QKQQ/m/qjgVWr7vAwAJ

**Comment by ocots** [comment]
[Consider using views for slices](https://docs.julialang.org/en/v1/manual/performance-tips/#man-performance-views)

**Reply by jbcaillau** [reply]
one could hope that the compiler, depending on what you do with the data, would take care of this

**Comment by jbcaillau** [comment]
@PierreMartinon @joseph-gergaud do we agree that the AD backend that is currently used in `CTDirect` for [ADNLPModels.jl](https://docs.juliahub.com/ADNLPModels/Emi8R/0.3.3/tutorial) is `ForwardDiff.jl` (the default, I assume)?

**Reply by jbcaillau** [reply]
Answer is yes (sorry for the noise). Might be worth to also try changing the AD backend to `Zygote.jl`
![IMG_2223](https://github.com/control-toolbox/control-toolbox.github.io/assets/62183989/ddc439d5-d8d8-4aca-98eb-a55fa0fa43c5)

**Comment by jbcaillau** [comment]
# [What is a “shared environment”? When should I use a “shared environment”?](https://discourse.julialang.org/t/what-is-a-shared-environment-when-should-i-use-a-shared-environment/100828)

🙏🏽 @rveltz

**Comment by jbcaillau** [comment]
#ad from https://discourse.julialang.org/t/this-month-in-julia-world-2023-07/102276 

FastDifferentiation.jl is a new type of hybrid automatic differentiation / symbolic differentiation tool. This one did not get much discussion, but it does seem like a tool that will be a gamechanger for many autodiff tasks. As always, do not forget that different tasks require different diff approaches, and check AbstractDifferentiation, ForwardDiff, TaylorDiff, Zygote, Enzyme, StochasticAD, and the early stage Diffractor.
The Diffractor 8 autodiff research project has a release with exciting new mostly-stable functionality.

**Comment by jbcaillau** [comment]
# Performance tips

Thanks @amontoison for [this documentation](https://jso.dev/NLPModelsIpopt.jl/previews/PR113/tips) on linear solvers + BLAS / LAPACK acceleration 👍🏽

**Comment by jbcaillau** [comment]
![IMG_2612](https://github.com/control-toolbox/control-toolbox.github.io/assets/62183989/17573191-fb88-4eb4-99c3-748f2d8fc7f7)
 
Greetings from Paris at [Julia and optimization days](https://julia-users-paris.github.io/workshop/en) with @ocots and @joseph-gergaud !

[Solving optimal control problems with Julia](https://control-toolbox.org/docs/optimalcontrol/stable/juliaopt2023.html)

**Comment by jbcaillau** [comment]
# Collaborative practices

[ColPrac](https://docs.sciml.ai/ColPrac/stable)

**Comment by rveltz** [comment]
and also https://domluna.github.io/JuliaFormatter.jl/dev/
https://domluna.github.io/JuliaFormatter.jl/dev/sciml_style/

**Comment by jbcaillau** [comment]
# TTL + TTFX

https://discourse.julialang.org/t/any-way-to-speed-up-loading-large-precompiled-packages/98762/35

**Comment by jbcaillau** [comment]
@amontoison @tmigot any comment on this?
https://www.nextplatform.com/2023/09/26/julia-still-not-grown-up-enough-to-ride-exascale-train
https://arxiv.org/pdf/2309.10292.pdf

**Comment by amontoison** [comment]
Hi @jbcaillau!
We are using the supercomputers Frontier and Aurora for our project at Argonne and I have a few comments for you:
- 50% gap compared to the vendor library is a great performance from my point of view.
- The gap is related to how the GPU backends are dealing with LLVM code. It was optimized for NVIDIA GPUs but not for AMD / Intel GPUs yet.
- the article just misses accessibility and development costs.

**Reply by jbcaillau** [reply]
@amontoison thanks for the feedback. sounds good, so. 

ps. on a related topic, check this [linkedin post](https://www.linkedin.com/posts/viralbshah_cudajl-51-unified-memory-and-cooperative-activity-7127680137436749824-0lXd?utm_source=share&utm_medium=member_ios)

**Reply by amontoison** [reply]
Thanks Jean-Baptiste!

**Comment by jbcaillau** [comment]
Anyone planning to attend?

https://juliacon.org/2024

**Reply by ocots** [reply]
I want to go. The conference will be in the stadium !

**Reply by jbcaillau** [reply]
@ocots seriously?

**Reply by ocots** [reply]
![Screenshot_20231121_194554_Opera.jpg](https://github.com/control-toolbox/control-toolbox.github.io/assets/66357348/0f35f36f-846d-475f-9be3-f2082c1d9929)

**Reply by ocots** [reply]
https://www.psv.nl/english-psv/club/philips-stadium.htm

**Comment by jbcaillau** [comment]
@rveltz anything more up to date / comprehensive on *Julia traits* than what is [here](https://github.com/orgs/control-toolbox/discussions/22#discussioncomment-5551423)?

**Comment by rveltz** [comment]
I like also the idea of https://github.com/jolin-io/WhereTraits.jl 
To me, the bigger problem is the one of interfaces when you develop such big package as yours.

**Reply by jbcaillau** [reply]
@rveltz Is it the disambiguation process (`@traits_order`) that you like in this package?

**Comment by jbcaillau** [comment]
@joseph-gergaud regarding an old discussion we had on Julia scoping, check [this](https://docs.julialang.org/en/v1/manual/variables-and-scoping/#man-scope-table).

**Comment by jbcaillau** [comment]
@ocots on `import` *vs.* `using`, check [this](https://web.mit.edu/julia_v0.6.2/julia/share/doc/julia/html/en/manual/modules.html#Summary-of-module-usage-1).

**Comment by ocots** [comment]
[Blue: a Style Guide for Julia](https://github.com/JuliaDiff/BlueStyle): This document specifies style conventions for Julia code. These conventions were created from a variety of sources including Python's [PEP8](http://legacy.python.org/dev/peps/pep-0008/), Julia's [Notes for Contributors](https://github.com/JuliaLang/julia/blob/master/CONTRIBUTING.md), and Julia's [Style Guide](https://docs.julialang.org/en/v1/manual/style-guide/).

**Comment by rveltz** [comment]
This is implemented in https://github.com/domluna/JuliaFormatter.jl

This can be a lot of refactoring to comply to a style depending on your project status… I was told it is the basic thing to get PR from external users

**Reply by jbcaillau** [reply]
thanks @rveltz (I remember you already pointed this style issue). For those on vscode [^1], these standards are taken into account in the current julia extension. Still interesting to try and run JuliaFormatter.jl on our code 🤞🏾...

[^1]: Romain check [this](https://sed-sam.gitlabpages.inria.fr/sed-sam/howto/remote-vscode-session.html) 🙂

**Comment by jbcaillau** [comment]
@ocots @joseph-gergaud on the [scope of loops](
https://m3g.github.io/JuliaNotes.jl/stable/loopscopes/#:~:text=Scope%20behavior%20of%20loops&text=Yet%2C%20in%20Julia%2C%20for%20performance,changed%20from%20outside%20the%20loop)

**Comment by jbcaillau** [comment]
On the program:
- Th. 14:00, "Dude, where's my code?" (T. Holy)

**Comment by jbcaillau** [comment]
@ocots @PierreMartinon @joseph-gergaud we need a place to save howto's (create a package, a doc, CI info, *etc.*) There is this discussion, this [wiki](https://github.com/control-toolbox/OptimalControl.jl/wiki), and quite a few things scattered under the control-toolbox organisation. any idea on a good way to do this properly?

**Reply by ocots** [reply]
It would be better [here](https://github.com/control-toolbox/control-toolbox.github.io/wiki).

**Comment by jbcaillau** [comment]
https://julialang.org/blog/2024/10/julia-1.11-highlights

**Reply by PierreMartinon** [reply]
Glad to see more performance improvements !

**Comment by amontoison** [comment]
https://github.com/LilithHafner/BasicAutoloads.jl

**Comment by ocots** [comment]
@jbcaillau https://juliareach.github.io

**Comment by jbcaillau** [comment]
@ocots @PierreMartinon @joseph-gergaud 
https://www.researchgate.net/profile/Francis-Gagnon-2/publication/385825845_ModelPredictiveControljl_advanced_process_control_made_easy_in_Julia/links/6736e2adf255d572866b6b4a/ModelPredictiveControljl-advanced-process-control-made-easy-in-Julia.pdf?origin=publication_detail&_tp=eyJjb250ZXh0Ijp7ImZpcnN0UGFnZSI6InB1YmxpY2F0aW9uIiwicGFnZSI6InB1YmxpY2F0aW9uRG93bmxvYWQiLCJwcmV2aW91c1BhZ2UiOiJwdWJsaWNhdGlvbiJ9fQ

**Comment by jbcaillau** [comment]
@ocots @PierreMartinon @joseph-gergaud take a look at https://github.com/JuliaParallel/Dagger.jl (parallel computation)

**Comment by jbcaillau** [comment]
anyone tested that https://github.com/JuliaArrays/OffsetArrays.jl ? @ocots @PierreMartinon @rveltz @amontoison

**Comment by jbcaillau** [comment]
@ocots 
- https://discourse.julialang.org/t/a-curated-list-of-gh-actions-that-improve-the-quality-of-your-package/134466/8
- https://github.com/JuliaBesties/BestieTemplate.jl
